include/openthread/platform/trel.h

```diff
@@ -135,6 +135,23 @@ typedef struct otPlatTrelPeerInfo
  */
 extern void otPlatTrelHandleDiscoveredPeerInfo(otInstance *aInstance, const otPlatTrelPeerInfo *aInfo);
 
+/**
+ * Notifies platform that a TREL packet is received from a peer using a different socket address than the one reported
+ * earlier from `otPlatTrelHandleDiscoveredPeerInfo()`.
+ *
+ * Ideally the platform underlying DNS-SD should detect changes to advertised port and addresses by peers, however,
+ * there are situations where this is not detected reliably. This function signals to the platform layer than we
+ * received a packet from a peer with it using a different port or address. This can be used by the playroom layer to
+ * restart/confirm the DNS-SD service/address resolution for the peer service and/or take any other relevant actions.
+ *
+ * @param[in] aInstance      The OpenThread instance.
+ * @param[in] aPeerSockAddr  The address of the peer, reported from `otPlatTrelHandleDiscoveredPeerInfo()` call.
+ * @param[in] aRxSockAddr    The address of received packet from the same peer (differs from @p aPeerSockAddr).
+ */
+void otPlatTrelNotifyPeerSocketAddressDifference(otInstance       *aInstance,
```

Comment:
Based on the latest discussion, there was consensus around just sending to the IPv6 Source Address of the last received message. Given that, do we need this API anymore?

Reply:
Yes. I think it is helpful to keep this functionality.

The main enhancement (the main idea in this PR) is that we can learn and update the peer address from the received TREL packet and use the same peer address for transmission as the one we last received. This was always the given assumption, at least for me. 🙂

Beyond that, notifying the platform about an address discrepancy can be helpful. The platform can, of course, choose to ignore this and take no action (which the current implementation does). But in the future, we may want to restart or refresh the DNS-SD address or service resolutions (e.g., resend queries) to ensure the peer's full address list can be discovered or refreshed.